tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/JMSXPathSelectorTest.java

```diff
@@ -0,0 +1,162 @@
+
+   @Test
+   public void testJMSSelectorsOpenWireProducerCoreConsumer() throws Exception {
+      testJMSSelectors(OpenWireConnection, CoreConnection);
+   }
+
+   @Test
+   public void testJMSSelectorsAMQPProducerOpenWireConsumer() throws Exception {
+      testJMSSelectors(AMQPConnection, OpenWireConnection);
+   }
+
+   @Test
+   public void testJMSSelectorsOpenWireProducerAMQPConsumer() throws Exception {
+      testJMSSelectors(OpenWireConnection, AMQPConnection);
+   }
+
+   public void testJMSSelectors(ConnectionSupplier producerConnectionSupplier, ConnectionSupplier consumerConnectionSupplier) throws Exception {
+      testJMSSelector(producerConnectionSupplier, consumerConnectionSupplier, NORMAL_QUEUE_NAME, "<root><a key='first' num='1'/><b key='second' num='2'>b</b></root>", "XPATH 'root/a'");
+   }
+
+   public void testJMSSelector(ConnectionSupplier producerConnectionSupplier, ConnectionSupplier consumerConnectionSupplier, String queueName, String body, String selector) throws Exception {
+      testJMSSelector(producerConnectionSupplier, consumerConnectionSupplier, queueName, body, selector, Message.DEFAULT_DELIVERY_MODE, Message.DEFAULT_PRIORITY, Message.DEFAULT_TIME_TO_LIVE);
+   }
+
+   public void testJMSSelector(ConnectionSupplier producerConnectionSupplier, ConnectionSupplier consumerConnectionSupplier, String queueName, String body, String selector, int deliveryMode, int priority, long timeToLive) throws Exception {
+      sendMessage(producerConnectionSupplier, queueName, body, deliveryMode, priority, timeToLive);
+      receive(consumerConnectionSupplier, queueName, body, selector);
+   }
+
+   private void receive(ConnectionSupplier consumerConnectionSupplier, String queueName, String body, String selector) throws JMSException {
+      try (Connection consumerConnection = consumerConnectionSupplier.createConnection()) {
+         Session consumerSession = consumerConnection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+         Queue consumerQueue = consumerSession.createQueue(queueName);
+         MessageConsumer consumer = consumerSession.createConsumer(consumerQueue, selector);
+         TextMessage msg = (TextMessage) consumer.receive(1000);
+         assertNotNull(msg);
+         assertEquals(body, msg.getText());
+         assertNull(consumer.receiveNoWait());
+         consumer.close();
+      }
+   }
+
+   private void sendMessage(ConnectionSupplier producerConnectionSupplier, String queueName, String body, int deliveryMode, int priority, long timeToLive) throws JMSException {
+      try (Connection producerConnection = producerConnectionSupplier.createConnection()) {
+         Session producerSession = producerConnection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+         Queue queue1 = producerSession.createQueue(queueName);
+         MessageProducer p = producerSession.createProducer(null);
+
+         TextMessage message1 = producerSession.createTextMessage();
+         message1.setText("hello");
```

Comment:
Fair enough.

Reply:
I actually meant sending it in addition to the clearly-not-xml body rather than instead of it, i.e one plain text, one partial-but-not-fully matching xml, and one matching xml body.  Kinf of like covering the edge and middle cases.